src/charm.py

```diff
@@ -443,6 +538,87 @@ def _on_create_identity_action(self, event: ActionEvent) -> None:
         event.log(f"Successfully created the identity: {res}")
         event.set_results({"identity-id": res})
 
+    def _on_run_migration_up_action(self, event: ActionEvent) -> None:
+        if not self.unit.is_leader():
+            event.fail("Non-leader unit cannot run migration action")
+            return
+
+        if not self._workload_service.is_running:
+            event.fail("Service is not ready. Please re-run the action when the charm is active")
+            return
+
+        if not peer_integration_exists(self):
+            event.fail("Peer integration is not ready yet")
+            return
+
+        event.log("Start migrating up the database")
+
+        timeout = float(event.params.get("timeout", 120))
+        try:
+            self._cli.migrate_up(
+                dsn=DatabaseConfig.load(self.database_requirer).dsn, timeout=timeout
```

Comment:
nitpick: same as before, can we instantiate it once and reuse the var? same on line 567, 592 and 602